data_steward/cdr_cleaner/clean_cdr_engine.py

```diff
@@ -37,129 +41,225 @@ def add_console_logging(add_handler):
 
 
-def clean_dataset(project=None, statements=None, data_stage=stage.UNSPECIFIED):
+def create_sandbox(project_id, dataset_id, client):
+    sandbox_dataset_id = sandbox.get_sandbox_dataset_id(dataset_id)
+    labels = {'type': 'sandbox', 'dataset_id': dataset_id}
+    sandbox_dataset = bq.define_dataset(project_id,
+                                        sandbox_dataset_id,
+                                        description=f'Sandbox for {dataset_id}',
+                                        label_or_tag=labels)
+    sandbox_dataset = client.create_dataset(sandbox_dataset, exists_ok=True)
+    return sandbox_dataset.dataset_id
+
+
+def clean_dataset(project_id=None, dataset_id=None, rules=None, **kwargs):
+    """
+    Run the assigned cleaning rules and return list of BQ job objects
+
+    :param project_id: identifies the project
+    :param dataset_id: identifies the dataset to clean
+    :param rules: a list of cleaning rule objects/functions as tuples
+    :param kwargs: keyword arguments a cleaning rule may require
+    :return all_jobs: List of BigQuery job objects
     """
-       Run the assigned cleaning rules.
-
-       :param project:  the project name
-       :param statements:  a list of dictionary objects to run the query
-       :param data_stage:  an enum to indicate what stage of the cleaning this is
-       """
-    if project is None or project == '' or project.isspace():
-        project = app_identity.get_application_id()
-        LOGGER.info('Project name not provided.  Using default.')
-
-    if statements is None:
-        statements = []
-
-    failures = 0
-    successes = 0
-
-    statement_length = len(statements)
-    for index, statement in enumerate(statements):
-        rule_query = statement.get(cdr_consts.QUERY, '')
-        legacy_sql = statement.get(cdr_consts.LEGACY_SQL, False)
-        destination_table = statement.get(cdr_consts.DESTINATION_TABLE, None)
-        retry = statement.get(cdr_consts.RETRY_COUNT,
-                              bq_consts.BQ_DEFAULT_RETRY_COUNT)
-        disposition = statement.get(cdr_consts.DISPOSITION,
-                                    bq_consts.WRITE_EMPTY)
-        destination_dataset = statement.get(cdr_consts.DESTINATION_DATASET,
-                                            None)
-        batch = statement.get(cdr_consts.BATCH, None)
-
-        module_name = statement.get(cdr_consts.MODULE_NAME,
-                                    cdr_consts.MODULE_NAME_DEFAULT_VALUE)
-        function_name = statement.get(cdr_consts.FUNCTION_NAME,
-                                      cdr_consts.FUNCTION_NAME_DEFAULT_VALUE)
+    if project_id is None or project_id == '' or project_id.isspace():
+        project_id = app_identity.get_application_id()
+        LOGGER.info(f"project_id not provided, using default {project_id}")
+
+    # Set up
+    client = bq.get_client(project_id=project_id)
+    sandbox_dataset_id = create_sandbox(project_id, dataset_id, client)
```

Comment:
Data hiding the sandbox like this seems to add complexity which _I think_ would otherwise be avoided if it were a `clean_dataset()` argument.

The behavior of `clean_dataset()` when the sandbox dataset already exists is ambiguous. If the sandbox is implicitly determined by `clean_dataset()` then the function will either A) fail or B) make the decision to proceed and overwrite (or whatever- some decision in which the caller is given no say). If `sandbox_dataset_id` is an argument, perhaps the same assumptions made about other args (`project_id`, `dataset_id`) would apply- the resource exists and this process has appropriate access.

Another angle is we can say the signature of `clean_dataset()` consists of three types of args:
1. the cleaning rules to apply (`rules`)
2. the params required by all cleaning rules (`project_id`, `dataset_id`,)
3. the params which may or may not be used by some of the cleaning rules (`kwargs`)

Looking at it this way it would seem that `sandbox_dataset_id` is missing from 2. Now some of the legacy rules do not accept `sandbox_dataset_id` but I understand this is just a technicality resulting from their being authored prior to our sandboxing practices. They could (if not "should") be refactored to use a sandbox.

I think making `sandbox_dataset_id` a parameter would also increase testability. I recall there was a test we looked at last week that needed implicit knowledge of what the `sandbox_dataset_id` was.

Reply:
Sure Mark, I was debating that as well. I thought of including it in `kwargs` earlier but a separate parameter is better. Do you think the sandbox should still be created by the engine?